drivers/i2c/i2c_ll_stm32.c

```diff
@@ -245,96 +245,78 @@ static int i2c_stm32_init(struct device *dev)
-};
-
-static struct i2c_stm32_data i2c_stm32_dev_data_1;
-
-DEVICE_AND_API_INIT(i2c_stm32_1, CONFIG_I2C_1_NAME, &i2c_stm32_init,
-		    &i2c_stm32_dev_data_1, &i2c_stm32_cfg_1,
-		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
-		    &api_funcs);
 
-#ifdef CONFIG_I2C_STM32_INTERRUPT
-static void i2c_stm32_irq_config_func_1(struct device *dev)
-{
 #ifdef CONFIG_I2C_STM32_COMBINED_INTERRUPT
-	IRQ_CONNECT(DT_I2C_1_COMBINED_IRQ, DT_I2C_1_COMBINED_IRQ_PRI,
-		   stm32_i2c_combined_isr, DEVICE_GET(i2c_stm32_1), 0);
-	irq_enable(DT_I2C_1_COMBINED_IRQ);
+#define STM32_I2C_IRQ_CONNECT_AND_ENABLE(name)				\
+	do {								\
+		IRQ_CONNECT(DT_##name##_COMBINED_IRQ,			\
+			    DT_##name##_COMBINED_IRQ_PRI,		\
+			    stm32_i2c_combined_isr,			\
+			    DEVICE_GET(i2c_stm32_##name), 0);		\
+		irq_enable(DT_##name##_COMBINED_IRQ);			\
+	} while (0)
 #else
-	IRQ_CONNECT(DT_I2C_1_EVENT_IRQ, DT_I2C_1_EVENT_IRQ_PRI,
-		   stm32_i2c_event_isr, DEVICE_GET(i2c_stm32_1), 0);
-	irq_enable(DT_I2C_1_EVENT_IRQ);
-
-	IRQ_CONNECT(DT_I2C_1_ERROR_IRQ, DT_I2C_1_ERROR_IRQ_PRI,
-		   stm32_i2c_error_isr, DEVICE_GET(i2c_stm32_1), 0);
-	irq_enable(DT_I2C_1_ERROR_IRQ);
+#define STM32_I2C_IRQ_CONNECT_AND_ENABLE(name)				\
+	do {								\
+		IRQ_CONNECT(DT_##name##_EVENT_IRQ,			\
+			    DT_##name##_EVENT_IRQ_PRI,			\
+			    stm32_i2c_event_isr,			\
+			    DEVICE_GET(i2c_stm32_##name), 0);		\
+		irq_enable(DT_##name##_EVENT_IRQ);			\
+									\
+		IRQ_CONNECT(DT_##name##_ERROR_IRQ,			\
+			    DT_##name##_ERROR_IRQ_PRI,			\
+			    stm32_i2c_error_isr,			\
+			    DEVICE_GET(i2c_stm32_##name), 0);		\
+		irq_enable(DT_##name##_ERROR_IRQ);			\
+	} while (0)
 #endif
+
+#define STM32_I2C_IRQ_HANDLER_DECL(name)				\
+static void i2c_stm32_irq_config_func_##name(struct device *dev)
+#define STM32_I2C_IRQ_HANDLER_FUNCTION(name)				\
+	.irq_config_func = i2c_stm32_irq_config_func_##name,
+#define STM32_I2C_IRQ_HANDLER(name)					\
+static void i2c_stm32_irq_config_func_##name(struct device *dev)	\
+{									\
+	STM32_I2C_IRQ_CONNECT_AND_ENABLE(name);				\
 }
+#else
+#define STM32_I2C_IRQ_HANDLER_DECL(name)
+#define STM32_I2C_IRQ_HANDLER_FUNCTION(name)
+#define STM32_I2C_IRQ_HANDLER(name)
 #endif
 
+#define STM32_I2C_INIT(name)						\
+STM32_I2C_IRQ_HANDLER_DECL(name);					\
+									\
+static const struct i2c_stm32_config i2c_stm32_cfg_##name = {		\
+	.i2c = (I2C_TypeDef *)DT_##name##_BASE_ADDRESS,			\
+	.pclken = {							\
+		.enr = DT_##name##_CLOCK_BITS,				\
+		.bus = DT_##name##_CLOCK_BUS,				\
+	},								\
+	STM32_I2C_IRQ_HANDLER_FUNCTION(name)				\
+	.bitrate = DT_##name##_BITRATE,					\
+};									\
+									\
+static struct i2c_stm32_data i2c_stm32_dev_data_##name;			\
+									\
+DEVICE_AND_API_INIT(i2c_stm32_##name, CONFIG_##name##_NAME,		\
+		    &i2c_stm32_init, &i2c_stm32_dev_data_##name,	\
+		    &i2c_stm32_cfg_##name,				\
+		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
+		    &api_funcs);					\
+									\
+STM32_I2C_IRQ_HANDLER(name)
+
```

Comment:
To enhance readability, what about adding a comment here, to clearly separate macro declaration and instances declaration part?